Implements Direct State Access and Multi Bind

mpnico opened the pull request:
### Implements OpenGL 4.4 Multi Bind and OpenGL 4.5 DSA
The goal of this PR is to reduce GL calls by using OpenGL 4.4 and 4.5 new extensions (`ARB_multi_bind` and `ARB_direct_state_access`), closing #2508.

Multi bind allows to bind multiple buffers in one call whereas DSA allows to change states of objects without binding them (buffers / textures, framebuffers, …).

Unfortunately, I didn't see major performance increase during my tests, do not expect more than a few %.

About VertexArray: the current implementation is not 100% full DSA. The "DSA" version of `glEnableVertexAttribArray` / `glDisableVertexAttribArray` are not used because of a big negative performance impact (around 15% on SMO).

About Framebuffer: as stated by @ReinUsesLisp in #2508, I encountered a bug with `CreateFramebuffers` so I made an exception only for NVidia.

### What to test
Basically every games for potential regressions. My system is NVidia only so AMD/Intel tests are needed. 

During my tests, I also encountered a strange flickering bug in Monster Hunter Rise which disappears when activating then disabling OpenGL log level. I assume it is a driver bug with my system but this game should be tested.

A-w-x commented:
intel linux

links awakening
[Ryujinx_1.0.0+be5dffc_2021-08-16_06-05-29.log](https://github.com/Ryujinx/Ryujinx/files/6992031/Ryujinx_1.0.0%2Bbe5dffc_2021-08-16_06-05-29.log)
![Screenshot_2021-08-16_06-09-52](https://user-images.githubusercontent.com/80461138/129557961-a764ab18-4b63-4f96-8319-2cf5ceb563b8.png)

riperiperi commented on Ryujinx.Graphics.OpenGL/Pipeline.cs:
```diff
@@ -1006,6 +994,7 @@ public void SetTransformFeedbackBuffers(ReadOnlySpan<BufferRange> buffers)
             }
 
             int count = Math.Min(buffers.Length, Constants.MaxTransformFeedbackBuffers);
+            int[] bufferHandles = new int[count];
```
Use `stackalloc` to avoid allocating arrays in the heap each time this is called.

riperiperi commented on Ryujinx.Graphics.OpenGL/Pipeline.cs:
```diff
@@ -1118,18 +1109,29 @@ private void SetBuffers(int first, ReadOnlySpan<BufferRange> buffers, bool isSto
 
+            int[] bufferHandles = new int[buffers.Length];
+            IntPtr[] bufferOffsets = new IntPtr[buffers.Length];
+            IntPtr[] bufferSizes = new IntPtr[buffers.Length];
```
`stackalloc` these - SetBuffers is called a lot and allocating arrays each time isn't great.

riperiperi commented on Ryujinx.Graphics.OpenGL/Framebuffer.cs:
```diff
@@ -17,9 +17,28 @@ class Framebuffer : IDisposable
+            {
+                GL.CreateFramebuffers(1, &localHandle);
+            }
+
```
Double newline.

riperiperi commented on Ryujinx.Graphics.OpenGL/Image/TextureView.cs:
```diff
@@ -276,8 +253,12 @@ private void ReadFrom2D(IntPtr data, int layer, int level, int width, int height
             TextureTarget target = Target.Convert();
 
             int mipSize = Info.GetMipSize2D(level);
+            int maxLevels = 1 + (int)Math.Floor(Math.Log2(width * height));
```
This calculation doesn't make any sense. Given a 256x256 texture, this will calculate 17 max levels... which is pretty far off.

```
256
128
64
32
16
8
4
2
1
```
There are 9 mip levels in this situation. With your current calculation, the correct result is obtained with 256x1.

mpnico replied:
Well this is a commit mistake. This was a test to avoid flood of OpenGL errors in logs in SMO which happens on master too. No relation with DSA so I will remove it.

riperiperi commented on Ryujinx.Graphics.OpenGL/VertexArray.cs:
```diff
@@ -20,9 +20,18 @@ class VertexArray : IDisposable
 
+        public unsafe int Create()
+        {
+            int localHnadle;
```
`localHandle`

riperiperi reviewed (commented):
Just a quick initial review. I've yet to do more testing, but I can confirm the Link's Awakening issue also happens on NVIDIA.
![image](https://user-images.githubusercontent.com/6294155/129566237-767117a0-faf6-4385-91f8-37cb393f300c.png)

mpnico commented:
Fixed Link's Awakening and applied code review changes.

mpnico commented:
I found a workaround for MHR flickering. I also pushed a few more fixes

gdkchan commented on Ryujinx.Graphics.OpenGL/Buffer.cs:
```diff
@@ -8,21 +8,18 @@ static class Buffer
+                size,
+                PixelFormat.RgbaInteger,
+                PixelType.UnsignedByte,
+                valueArr.ToArray());
```
Calling `ToArray` here negates the benefits of using the stackalloc (as you will allocate a new array on the heap anyway).

mpnico replied:
This is the fix for Monster Hunter Rise. Maybe using `MemoryMarshal.GetReference` is better ?

gdkchan replied:
If theres an overload that takes ref, you can just pass `ref valueArr[0]`. Or well, just `ref value` since theres only one value.

mpnico replied:
Method support to pass `ref value` but does not fix the bug that way. `MemoryMarshal.GetReference` does fix it but I don't know it efficiency.

gdkchan replied:
I think passing the array is just fixing the issue by chance. It's possible that the driver is reading past the pointer that we pass. Try changing the stackalloc size from 1 to 4, and passing the pointer directly, see if that fixes it.

mpnico replied:
Good idea but unfortunately it doesn't fix the bug. But you gave me the idea to change `PixelInternalFormat` to a 1 component only (`GL_R32UI`) format and it fixes the bug. I don't know if it's very less efficient though.

mpnico replied:
Finally the best fix is using `MemoryMarshal.GetReference` or not using DSA here. I choose the first one.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/Texture.cs:
```diff
@@ -833,7 +833,7 @@ public void Flush(bool tracked = true)
         public void ExternalFlush(ulong address, ulong size)
         {
-            if (!IsModified)
+            if (!IsModified || HostTexture.isDisposed())
```
There is no need for this `isDisposed` method (btw the correct name would be `IsDisposed` as funciton names always start with upper case). You can just have a bool field and set it to true when `DisposeTextures` is called. Then here it would just check `if (!IsModified || _disposed)`.

gdkchan replied:
Also I'm not sure of the effects of leaving `IsModified` set to true. Maybe it could cause the texture to be flushed all the time. @riperiperi should be able to give a better suggestion here.

mpnico replied:
I did some tests and, at least in MK8D where I found those disposed textures, and no excessive flush happen so I will make the well advised changes and won't touch `IsModified`

mpnico commented:
Rebased and fixed a problem with some UE games. It need some more tests on AMD as I also pushed a potential fix for garbage pixels at edges of the window.

TSRBerry commented:
Needs a rebase

marysaka commented:
Closing as stale.
